contrib/src/test/scala/akka/stream/contrib/UnfoldFlowSpec.scala

```diff
@@ -0,0 +1,198 @@
+            sink.ensureSubscription()
+            killSwitch.abort(kill)
+            sink.expectError(kill)
+          }
+
+          "fail when inner stream is canceled and pulled before completion" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            val kill = new Exception("KILL!")
+            sink.ensureSubscription()
+            killSwitch.abort(kill)
+            sink.request(1)
+            sink.expectError(kill)
+          }
+
+          "fail after 3 elements when aborted" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            val kill = new Exception("KILL!")
+            sink.requestNext(1)
+            sink.requestNext(2)
+            sink.requestNext(3)
+            killSwitch.abort(kill)
+            sink.expectError(kill)
+          }
+
+          "complete gracefully when stopped" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            sink.ensureSubscription()
+            killSwitch.shutdown()
+            sink.expectComplete()
+          }
+
+          "complete gracefully after 3 elements when stopped" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            sink.requestNext(1)
+            sink.requestNext(2)
+            sink.requestNext(3)
+            killSwitch.shutdown()
+            sink.expectComplete()
+          }
+        }
+
+        "killSwitch appended to flow" should {
+
+          val source = SourceGen.unfoldFlow(1)(Flow[Int].map { n =>
+            (n + 1, n)
+          }.viaMat(KillSwitches.single)(Keep.right))
+
+          "fail instantly when aborted" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            val kill = new Exception("KILL!")
+            sink.ensureSubscription()
+            killSwitch.abort(kill)
+            sink.expectError(kill)
+          }
+
+          "fail after 3 elements when aborted" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            val kill = new Exception("KILL!")
+            sink.requestNext(1)
+            sink.requestNext(2)
+            sink.requestNext(3)
+            killSwitch.abort(kill)
+            sink.expectError(kill)
+          }
+
+          "complete gracefully when stopped" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            sink.ensureSubscription()
+            killSwitch.shutdown()
+            sink.expectComplete()
+          }
+
+          "complete gracefully after 3 elements when stopped" in {
+            val (killSwitch, sink) = source.toMat(TestSink.probe)(Keep.both).run()
+            sink.requestNext(1)
+            sink.requestNext(2)
+            sink.requestNext(3)
+            killSwitch.shutdown()
+            sink.expectComplete()
+
+          }
+        }
+      }
+
```

Comment:
Please also add a test case where the inner Flow has a buffer (i.e. just prepend or append a buffer stage). That exercises more paths.